test/metabase/pulse_test.clj

```diff
@@ -105,27 +103,31 @@
           :when        f]
     (assert (fn? f))
     (testing (format "sent to %s channel" channel-type)
-      (testing (when (= :email channel-type) @mt/inbox)
-        (mt/with-temp [Card          {card-id :id} (merge {:name    pulse.test-util/card-name
-                                                           :display (or display :line)}
-                                                          card)]
-          (with-pulse-for-card [{pulse-id :id}
-                                {:card       card-id
-                                 :pulse      pulse
-                                 :pulse-card pulse-card
-                                 :channel    channel-type}]
-            (letfn [(thunk* []
-                      (f {:card-id card-id, :pulse-id pulse-id}
-                         (metabase.pulse/send-pulse! (pulse/retrieve-notification pulse-id))))
-                    (thunk []
-                      (if fixture
-                        (fixture {:card-id card-id, :pulse-id pulse-id} thunk*)
-                        (thunk*)))]
-              (case channel-type
-                :email (pulse.test-util/email-test-setup (thunk))
-                :slack (pulse.test-util/slack-test-setup (thunk))))))))))
-
-(defn- tests
+      (mt/with-temp [Card          {card-id :id} (merge {:name    pulse.test-util/card-name
+                                                         :display (or display :line)}
+                                                        card)]
+        (with-pulse-for-card [{pulse-id :id}
+                              {:card       card-id
+                               :pulse      pulse
+                               :pulse-card pulse-card
+                               :channel    channel-type}]
+          (letfn [(thunk* []
+                    (f {:card-id card-id, :pulse-id pulse-id}
+                       ((if (= :email channel-type)
+                          :channel/email
+                          :channel/slack)
+                        (pulse.test-util/with-captured-channel-send-messages!
```

Comment:
if you look at how tests were written before you'll see 2 different approaches with email and slack.

For slack, the assertions are made against the result of `(send-pulse!)`, this is done by having a bunch of redefs to make `(send-pulse!)` return the rendered-notification for slack only.

But for email, the test ignore the result of `(send-pulse!)`, instead it checks results from the global `(mt/inbox)` atom.

I also think the recent flakes with pulse is because we were using this global `inbox` atom, it could be that parallel tests mutate the inbox while a test is running.

So I create a new `pulse.test-util/with-captured-channel-send-messages!` macro which will capture all calls to `(channel/send!)`, this function should take a `message` that is already rendered and is ready to send. Which means: for email, it's a html with recipients info, for slack: It contains all the rendered attachments and text.

Reply:
I also believe this will fix our flakes with pulse recently.